django/db/migrations/executor.py

```diff
@@ -71,46 +72,91 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
         """
         if plan is None:
             plan = self.migration_plan(targets)
-        migrations_to_run = {m[0] for m in plan}
         # Create the forwards plan Django would follow on an empty database
         full_plan = self.migration_plan(self.loader.graph.leaf_nodes(), clean_start=True)
-        # Holds all states right before a migration is applied
-        # if the migration is being run.
+
+        all_forwards = all(not backwards for mig, backwards in plan)
+        all_backwards = all(backwards for mig, backwards in plan)
+
+        if all_forwards and not all_backwards:
+            self._migrate_all_forwards(plan, full_plan, fake=fake, fake_initial=fake_initial)
+        elif not all_forwards and all_backwards:
+            self._migrate_all_backwards(plan, full_plan, fake=fake)
+        elif plan:
```

Comment:
I don't like how this entire if block looks. Is `plan` allowed to be `None` considering the check/assignment on line 73 and 74? I'd prefer to call out that it can not be forwards and backwards by doing something like:

```
if all_forwards == all_backards: raise..  # this may not be correct if `plan` can be `None`
elif all_forwards: migrate_forwards()
else: migrate_backwards()
```

I feel this highlights the invariant, and avoids potential problems of misreading the bool combinations in the original code.